crates/rust/src/lib.rs

```diff
@@ -212,24 +254,45 @@ impl WorldGenerator for RustWasm {
         name: &WorldKey,
         id: InterfaceId,
         _files: &mut Files,
-    ) {
-        let mut gen = self.interface(None, resolve, false);
-        gen.current_interface = Some((id, name));
-        gen.types(id);
-        let trait_name = match name {
-            WorldKey::Name(name) => name.to_upper_camel_case(),
-            WorldKey::Interface(id) => resolve.interfaces[*id]
-                .name
-                .as_ref()
-                .unwrap()
-                .to_upper_camel_case(),
+    ) -> Result<()> {
+        let (pkg, inner_name) = match name {
+            WorldKey::Name(name) => (None, name),
+            WorldKey::Interface(id) => {
+                let interface = &resolve.interfaces[*id];
+                (
+                    Some(&resolve.packages[interface.package.unwrap()].name),
+                    interface.name.as_ref().unwrap(),
+                )
+            }
         };
+        let path = format!(
+            "{}{inner_name}",
+            if let Some(pkg) = pkg {
+                format!("{}::{}::", pkg.namespace, pkg.name)
+            } else {
+                String::new()
+            }
+        );
+        let impl_name = self
+            .opts
+            .interface_exports
+            .get(&path)
+            .cloned()
+            .or_else(|| self.opts.stubs.then(|| "Stub".to_owned()))
+            .ok_or_else(|| format!("`interface_exports` parameter required for `{path}`"));
+        let mut gen = self.interface(Identifier::Interface(id, name), None, resolve, false);
+        let (snake, path_to_root, pkg) = gen.start_append_submodule(name);
+        gen.types(id);
         gen.generate_exports(
-            &trait_name,
+            &inner_name.to_upper_camel_case(),
+            Some(&path),
+            impl_name.as_deref(),
             Some(name),
             resolve.interfaces[id].functions.values(),
-        );
-        gen.finish_append_submodule(name);
+        )
+        .map_err(|e| anyhow!("{e}"))?;
```

Comment:
Could `generate_exports` change to return `Result<T>` to avoid the need for this?